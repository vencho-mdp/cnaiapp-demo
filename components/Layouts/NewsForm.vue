<template>
  <form
    class="flex flex-col flex-grow mx-4 justify-evenly"
    @submit.prevent="add()"
  >
    <div class="flex flex-col justify-between my-8">
      <v-label class="mb-2 !text-white"> TÃ­tulo </v-label>
      <!-- v-model no funciona con v-on $listeners
      https://github.com/vuejs/vue/issues/7042
       -->
      <v-text-input
        data-test="title"
        class="w-full"
        :value="form.title"
        @input.native="form.title = $event.target.value"
      />
    </div>
    <div class="flex flex-col justify-between my-8">
      <v-label class="mb-2 !text-white"> Contenido </v-label>
      <client-only>
        <vue-editor
          v-model="form.content"
          :editor-toolbar="customToolbar"
          class="rounded-xl"
          data-test="content"
        />
      </client-only>
    </div>
    <file-input
      :image.sync="form.image"
      data-test="file_uploader"
      :form-data="formData"
      class="mt-16"
    />
    <form-buttons
      :is-add-button-invalid="$v.$invalid"
      class="my-12"
      @handleCancelButtonClick="$emit('closeSidebar', 'cancel')"
    />
  </form>
</template>

<script>
import { required } from "vuelidate/lib/validators";

export default {
  props: {
    formData: {
      type: Object,
      default: () => null,
    },
  },
  data() {
    return {
      form: this.formData || {
        content: undefined,
        title: undefined,
        image: undefined,
      },
      customToolbar: [
        ["bold", "italic", "underline"],
        [{ list: "ordered" }, { list: "bullet" }],
      ],
    };
  },
  validations: {
    form: {
      title: {
        required,
      },
      content: {
        required,
      },
    },
  },
  methods: {
    async add() {
      // Sacar atributos del componente card
      const { label, desc, ...data } = this.form;
      const form = new FormData();

      form.append("title", data.title);
      form.append("content", data.content);

      if (data.image) {
        form.append("image", this.form.image);
      }

      if (this.formData?.id) {
        form.append("id", this.formData.id);
      }

      if (this.formData?.original_image_extension) {
        form.append(
          "original_image_extension",
          this.formData.original_image_extension
        );
      }

      try {
        await this.$axios[this.formData ? "$put" : "$post"]("/api/news", form);
        this.$emit("closeSidebar");
      } catch (error) {
        this.$reportNetworkError(error);
      }
    },
  },
};
</script>

<style>
.image-shadow {
  box-shadow: inset 0 0 0 1000px rgba(0, 0, 0, 0.2);
}

.ql-editor {
  min-height: 48px !important;
  @apply bg-white rounded-b-xl max-h-40;
}
.ql-toolbar {
  @apply bg-white rounded-t-xl;
}
.ql-editor p,
.ql-editor strong,
.ql-editor u,
.ql-editor em,
.ql-editor li {
  @apply !text-black;
}
#quill-container {
  @apply !border-none;
}
</style>
