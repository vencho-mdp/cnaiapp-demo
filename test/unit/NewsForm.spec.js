import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuelidate from 'vuelidate'
import NewsForm from '@/components/Layouts/NewsForm.vue'

const localVue = createLocalVue()
localVue.use(Vuelidate)

describe('NewsForm', () => {
  const factory = (values = {}) => {
    return shallowMount(NewsForm, {
      propsData: values,
      localVue
    })
  }
  it('renders correctly', () => {
    const wrapper = factory()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('does not send form if it is invalid', () => {
    const wrapper = factory()
    expect(wrapper.vm.$v.$invalid).toBe(true)
  })
  it('sends form if it is valid', () => {
    const wrapper = factory()
    wrapper.setData({
      form: {
        title: 'test',
        content: 'test',
        image: 'test'
      }
    })
    expect(wrapper.vm.$v.$invalid).toBe(false)
  })
})
