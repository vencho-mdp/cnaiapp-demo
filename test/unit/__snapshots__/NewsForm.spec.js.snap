// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsForm renders correctly 1`] = `
"<form class=\\"flex flex-col flex-grow mx-4 justify-evenly\\">
  <div class=\\"flex flex-col justify-between my-8\\">
    <v-label-stub class=\\"mb-2 !text-white\\">
      TÃ­tulo
    </v-label-stub>
    <v-text-input-stub data-test=\\"title\\" class=\\"w-full\\"></v-text-input-stub>
  </div>
  <div class=\\"flex flex-col justify-between my-8\\">
    <v-label-stub class=\\"mb-2 !text-white\\">
      Contenido
    </v-label-stub> <span><vue-editor editor-toolbar=\\"bold,italic,underline,[object Object],[object Object]\\" data-test=\\"content\\" class=\\"rounded-xl\\"></vue-editor></span>
  </div>
  <file-input-stub data-test=\\"file_uploader\\" class=\\"mt-16\\"></file-input-stub>
  <form-buttons-stub isaddbuttoninvalid=\\"true\\" addbuttontext=\\"Guardar\\" class=\\"my-12\\"></form-buttons-stub>
</form>"
`;
